src/dmd/identifier.d

```diff
@@ -66,6 +68,17 @@ public:
         this(name[0 .. strlen(name)], TOK.identifier);
     }
 
+    static void initialize()
+    {
+        anonymous_ = new Identifier("__anonymous", TOK.identifier);
+    }
+
+    /// Placeholder for an anonymous identifier.
```

Comment:
It isn't a placeholder. It's a singleton. The code could be simplified and made clearer by moving the initialization of `anonymous_` and the `__gshared` variable into this function and making it initialize lazilly.

Reply:
That would cause potential issues with threads. But if we don't care about that I can make the change.

Reply:
What Walter is saying is this should either be a `generateId` and thus be local, or it should be part of [`Id`](https://github.com/dlang/dmd/blob/master/src/dmd/id.d#L26) like all the other compiler recognised/generated ones.

Reply:
The front end is single threaded so no issue with threads. It could be made multithreaded by using `static` instead of `__gshared`, but there's no reason to do that at this time.

Reply:
> What Walter is saying is this should either be a generateId and thus be local, or it should be part of Id like all the other compiler recognised/generated ones.

It is generated locally. It's intended usage is like this:

```d
void foo(Identifier ident)
{
    if (ident is Identifier.anonymous) // hence why it's a placeholder
        ident = new Identifier(...);
}
```

Reply:
So its a sentinel. Can you get away with using `null`?

Reply:
> So its a sentinel.

That might be a better name.

> Can you get away with using null?

Yes, but:

* It's not descriptive/clear what's going on. I prefer encoding the intent in the actual code than in comments
* There's nothing stopping `ident` from being `null` for other reasons

Reply:
>There's nothing stopping ident from being null for other reasons

I was hoping that was not going to be the case, oh well.

Reply:
@thewilsonator another approach than using a sentinel would be to add a new field to `Identifier` to indicate it's anonymous. This would have the advantage of allowing to add a more descriptive name to an identifier that can help with debugging.